llvm/lib/Target/NVPTX/NVPTXAliasAnalysis.cpp

```diff
@@ -115,3 +116,30 @@ ModRefInfo NVPTXAAResult::getModRefInfoMask(const MemoryLocation &Loc,
 
   return ModRefInfo::ModRef;
 }
+
+MemoryEffects NVPTXAAResult::getMemoryEffects(const CallBase *Call,
+                                              AAQueryInfo &AAQI) {
+  // Inline assembly with no side-effect or memory clobbers should not
+  // indirectly access memory in the PTX specification.
+  if (const auto *IA = dyn_cast<InlineAsm>(Call->getCalledOperand())) {
+    // Volatile is translated as side-effects.
+    if (IA->hasSideEffects())
+      return MemoryEffects::unknown();
+
+    for (const InlineAsm::ConstraintInfo &Constraint : IA->ParseConstraints()) {
+      // Indirect constraints (e.g. =*m) are unsupported in inline PTX.
+      if (Constraint.isIndirect)
+        return MemoryEffects::unknown();
+
+      // Memory clobbers prevent optimization.
+      if (!(Constraint.Type & InlineAsm::ConstraintPrefix::isClobber))
+        continue;
+      for (const std::string &Code : Constraint.Codes)
+        if (Code == "{memory}")
+          return MemoryEffects::unknown();
```

Comment:
Nit: `if (isClobber && any_of (Codes, [](const auto &Code){ Code == "{memory}"}) ` may be a bit more readable.